### PR TITLE
expression,types: Adjusts UNIX_TIMESTAMP() for non-existing DST values (#28739)

### LINKED ISSUE #28739: expression: Non-existing local timestamps during DST spring forward got converted to 0 instead of getting the transition timestamp
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
set time_zone = 'Europe/Vilnius';
select unix_timestamp('2020-03-29 03:45:00');
```

### 2. What did you expect to see? (Required)

On MySQL 8.0, after the server time zone got [populated](https://dev.mysql.com/doc/refman/8.0/en/time-zone-support.html#time-zone-installation):

```sql
mysql> select unix_timestamp('2020-03-29 03:45:00');
+---------------------------------------+
| unix_timestamp('2020-03-29 03:45:00') |
+---------------------------------------+
|                            1585443600 |
+---------------------------------------+
1 row in set (0.00 sec)
```

(Note: 1,585,443,600 = 2020-03-29 01:00:00 UTC, the instant when the spring forward happens)

### 3. What did you see instead (Required)

```
mysql> select unix_timestamp('2020-03-29 03:45:00');
+---------------------------------------+
| unix_timestamp('2020-03-29 03:45:00') |
+---------------------------------------+
|                                     0 |
+---------------------------------------+
1 row in set (0.15 sec)
```

### 4. What is your TiDB version? (Required)

```
Release Version: v5.2.1
Edition: Community
Git Commit Hash: cd8fb24c5f7ebd9d479ed228bb41848bd5e97445
Git Branch: heads/refs/tags/v5.2.1
UTC Build Time: 2021-09-08 02:32:56
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #28739

Problem Summary
TiDB did use the standard go time library which would interpret the time in a more relaxed and simple manner, resulting in interpreting the time one hour forward, instead of truncating it to the the first valid time during a Daylight Saving Time after the gap in local time.

### What is changed and how it works?
Unfortunately there are no good go time library function to use, so first part of the implementation takes the optimised approach and guessing that the non-existing time occurred on a DST shift, increasing the offset on a full hour.
Then a fallback with a binary search takes care of non-common cases like 'Australia/Lord_Howe' time zones.

### Check List

Tests 

- [X] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [X] Changes MySQL compatibility

### Release note

```release-note
UNIX_TIMESTAMP(datetime) would return 0 for non-existing date time values, which occurred in the Daylight Saving Time transition that increases the time zone offset, i.e. moving from normal time to summer time.
```

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -4856,7 +4856,7 @@ func (b *builtinUnixTimestampIntSig) evalIntWithCtx(ctx sessionctx.Context, row
 	}
 
 	tz := ctx.GetSessionVars().Location()
-	t, err := val.GoTime(tz)
+	t, err := val.AdjustedGoTime(tz)
```
Comment (morgo) [inline expression/builtin_time.go:4859]: I think expression/builtin_time_vec.go will need modifying too?

Comment (mjonss) [reply expression/builtin_time.go:4859]: Good catch! I added the fix there too.

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: types/core_time.go [hunk 2]
```diff
@@ -184,6 +184,59 @@ func (t CoreTime) GoTime(loc *gotime.Location) (gotime.Time, error) {
 	return tm, nil
 }
 
+// FindZoneTransition check for one Time Zone transition within +/- 4h
+// Currently the needed functions are not exported, if gotime.Location.lookup would be exported
+// then it would be easy to use that directly
+func FindZoneTransition(tIn gotime.Time, loc *gotime.Location) (gotime.Time, error) {
```
[the remark this answers is not in the document]
Comment (mjonss) [reply types/core_time.go]: Thanks for finding it! I just removed it and added some tests for fractional seconds as well.

**Review by xhebox — APPROVED** [review]

**Review by djshow832 — APPROVED** [review]

**Comment by djshow832** [comment]
There was an unstable test: https://github.com/pingcap/tidb/issues/30584

**Pull request merged**